### PR TITLE
Deprecate countnz in favor of using count(predicate, x)

### PR DESCRIPTION
**Pull request by andreasnoack**
As discussed in https://github.com/JuliaLang/julia/issues/23005, it is ambiguous which kind of non-zero to test with (`!iszero` or `t -> t != 0`) so with this PR the user is required to be explicit. To summarize the difference
```julia
julia> map(t -> t == 0, [rand(2) for i in 1:3])
3-element Array{Bool,1}:
 false
 false
 false

julia> map(!iszero, [rand(2) for i in 1:3])
3-element Array{Bool,1}:
 true
 true
 true

julia> map(t -> t != 0, ["Julia" for i in 1:3])
3-element Array{Bool,1}:
 true
 true
 true

julia> map(!iszero, ["Julia" for i in 1:3])
ERROR: MethodError: no method matching zero(::String)
Closest candidates are:
  zero(::Type{Base.LibGit2.GitHash}) at libgit2/oid.jl:166
  zero(::Type{Base.Pkg.Resolve.VersionWeights.VWPreBuildItem}) at pkg/resolve/versionweight.jl:80
  zero(::Type{Base.Pkg.Resolve.VersionWeights.VWPreBuild}) at pkg/resolve/versionweight.jl:122
  ...
Stacktrace:
 [1] iszero(::String) at ./number.jl:27
 [2] (::getfield(Base, Symbol("##61#62")){typeof(iszero)})(::String, ::Vararg{String,N} where N) at ./operators.jl:974
 [3] _collect at ./array.jl:678 [inlined]
 [4] collect_similar(::Array{String,1}, ::Base.Generator{Array{String,1},getfield(Base, Symbol("##61#62")){typeof(iszero)}}) at ./array.jl:625
 [5] map(::Function, ::Array{String,1}) at ./abstractarray.jl:1969
```
The deprecation suggests `!iszero` as well as `t -> t != 0` but still calls `t -> t != 0` to avoid breaking current behavior. We might want to revisit sparse string arrays which is the only thing I'm aware of that relies on this behavior but that is for a different PR.

### CODE REVIEW

**Review by mbauman — APPROVED** [review]

File: base/multidimensional.jl [hunk 1]
```diff
@@ -576,7 +576,7 @@ end
 
 @generated function findn(A::AbstractArray{T,N}) where {T,N}
     quote
-        nnzA = countnz(A)
+        nnzA = count(!iszero, A)
```
Comment (nalimilan) [inline base/multidimensional.jl]: This is inconsistent with the `findn` definition above and with its docstring, which states "determined by `A[i]!=0`".

File: base/reduce.jl [hunk 2]
```diff
@@ -703,22 +703,3 @@ function count(pred, a::AbstractArray)
...
-3×3 Array{Int64,2}:
- 1  2  4
- 0  0  1
- 1  1  0
-
-julia> countnz(A)
-6
-```
-"""
-countnz(a) = count(x -> x != 0, a)
```
Comment (nalimilan) [inline base/reduce.jl]: Note that the deprecation is different from the current definition, which (as you remarked) is technically breaking. I think the best approach would be to use `Base.depwarn` to print a custom deprecation warning proposing both forms, but continue calling `x -> x != 0` under the hood.

File: base/sparse/sparsematrix.jl [hunk 3]
```diff
@@ -70,9 +70,9 @@ julia> nnz(A)
 3
 ```
 """
-nnz(S::SparseMatrixCSC) = Int(S.colptr[S.n + 1]-1)
-countnz(S::SparseMatrixCSC) = countnz(S.nzval)
-count(S::SparseMatrixCSC) = count(S.nzval)
+nnz(S::SparseMatrixCSC)      = Int(S.colptr[S.n + 1] - 1)
+count(S::SparseMatrixCSC)    = count(S.nzval)
+count(f, S::SparseMatrixCSC) = count(f, S.nzval) + f(zero(eltype(S)))*(prod(size(S)) - nnz(S))
```
Comment (nalimilan) [inline base/sparse/sparsematrix.jl]: What's this new method? Should go in a different commit if that's really intended.

Comment (andreasnoack) [reply base/sparse/sparsematrix.jl]: This method was missing. There is already a (dense) generic fallback here https://github.com/JuliaLang/julia/blob/5645fb212b90910b06c5a06f08c746fdd5b82f80/base/reduce.jl#L698-L704 but not a sparse version. The `count(pred, x)` method is suggested by the deprecation warning so I think it makes sense to add it in this PR.

File: base/reduce.jl (hunk 2 above)
Comment (andreasnoack) [reply base/reduce.jl]: I've changed the deprecation to suggest both versions following your suggestion.

**Comment by nalimilan** [comment]
Unfortunately, using `x -> x != 0` in `findn` triggers this error: `generated function body is not pure. this likely means it contains a closure or comprehension.` Just create an equivalent function?

File: base/sparse/sparsematrix.jl (hunk 3 above)
Comment (nalimilan) [reply base/sparse/sparsematrix.jl]: Got it. Might be worth a word in the commit message.
